File: Plain Craft Launcher 2/Controls/Behaviors/LazyLoadBehavior.vb
Imports System.Runtime.CompilerServices
Imports Microsoft.Xaml.Behaviors

Module LazyLoader
    <Extension()>
    Public Sub EnableLazyLoad(element As FrameworkElement, action As Action)
        Dim behavior As New LazyLoadBehavior()
        behavior.Action = action
        Interaction.GetBehaviors(element).Add(behavior)
    End Sub
End Module

Public Class LazyLoadBehavior
    Inherits Behavior(Of FrameworkElement)

    Public Shared ReadOnly ActionProperty As DependencyProperty =
        DependencyProperty.Register(NameOf(Action), GetType(Action), GetType(LazyLoadBehavior), New PropertyMetadata(Nothing))

    Public Property Action As Action
        Get
            Return CType(GetValue(ActionProperty), Action)
        End Get
        Set(value As Action)
            SetValue(ActionProperty, value)
        End Set
    End Property

    Protected Overrides Sub OnAttached()
        MyBase.OnAttached()
        AddHandler AssociatedObject.LayoutUpdated, AddressOf OnLayoutUpdated
    End Sub

    Protected Overrides Sub OnDetaching()
        RemoveHandler AssociatedObject.LayoutUpdated, AddressOf OnLayoutUpdated
        MyBase.OnDetaching()
    End Sub

    Private Sub OnLayoutUpdated(sender As Object, e As EventArgs)
        If AssociatedObject.RenderSize.Width < Double.Epsilon Then Return
        If Not AssociatedObject.IsVisible Then Return

        Dim scrollViewer = FindParentScrollViewer(AssociatedObject)
        If scrollViewer Is Nothing Then Return

        Dim elementBounds = AssociatedObject.TransformToAncestor(scrollViewer).TransformBounds(
            New Rect(New Point(0, 0), AssociatedObject.RenderSize))
        Dim viewport = New Rect(0, 0, scrollViewer.ViewportWidth, scrollViewer.ViewportHeight)

        If viewport.IntersectsWith(elementBounds) Then
            Action?.Invoke()
            ' 仅执行一次
            RemoveHandler AssociatedObject.LayoutUpdated, AddressOf OnLayoutUpdated
        End If
    End Sub

    Private Function FindParentScrollViewer(d As DependencyObject) As ScrollViewer
        While d IsNot Nothing
            If TypeOf d Is ScrollViewer Then Return CType(d, ScrollViewer)
            d = VisualTreeHelper.GetParent(d)
        End While
        Return Nothing
    End Function
End Class

File: PCL-CE/Plain Craft Launcher 2/Controls/Behaviors/LazyLoadBehavior.cs
using System.Windows;
using System.Windows.Controls;
using System.Windows.Media;
using Microsoft.Xaml.Behaviors;

namespace PCL;

internal static class LazyLoader
{
    public static void EnableLazyLoad(this FrameworkElement element, Action action)
    {
        var behavior = new LazyLoadBehavior();
        behavior.Action = action;
        Interaction.GetBehaviors(element).Add(behavior);
    }
}

public class LazyLoadBehavior : Behavior<FrameworkElement>
{
    public static readonly DependencyProperty ActionProperty = DependencyProperty.Register(nameof(Action),
        typeof(Action), typeof(LazyLoadBehavior), new PropertyMetadata(null));

    public Action Action
    {
        get => (Action)GetValue(ActionProperty);
        set => SetValue(ActionProperty, value);
    }

    protected override void OnAttached()
    {
        base.OnAttached();
        AssociatedObject.LayoutUpdated += OnLayoutUpdated;
    }

    protected override void OnDetaching()
    {
        AssociatedObject.LayoutUpdated -= OnLayoutUpdated;
        base.OnDetaching();
    }

    private void OnLayoutUpdated(object sender, EventArgs e)
    {
        var element = AssociatedObject;
        if (element is null || element.RenderSize.Width < double.Epsilon || !element.IsVisible)
            return;

        var scrollViewer = FindParentScrollViewer(element);
        if (scrollViewer is null)
            return;

        var elementBounds = element.TransformToAncestor(scrollViewer)
            .TransformBounds(new Rect(new Point(0d, 0d), element.RenderSize));
        var viewport = new Rect(0d, 0d, scrollViewer.ViewportWidth, scrollViewer.ViewportHeight);

        if (!viewport.IntersectsWith(elementBounds))
            return;

        Action?.Invoke();
        // 仅执行一次
        element.LayoutUpdated -= OnLayoutUpdated;
    }

    private static ScrollViewer FindParentScrollViewer(DependencyObject d)
    {
        for (var current = d; current is not null; current = VisualTreeHelper.GetParent(current))
        {
            if (current is ScrollViewer scrollViewer)
                return scrollViewer;
        }

        return null;
    }
}
